instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,257 @@
+import logging
+from functools import wraps
+from typing import Callable, Dict, List, MutableMapping, Sequence, Type
+
+from sklearn.base import BaseEstimator
+from sklearn.pipeline import FeatureUnion, Pipeline
+
+from opentelemetry.instrumentation.sklearn.version import __version__
+from opentelemetry.trace import get_tracer
+
+logger = logging.getLogger(__name__)
+
+
+def implement_spans(func: Callable, estimator_name: str):
+    """Wrap the method call with a span.
+
+    Args:
+        func: A callable to be wrapped in a span
+        estimator_name: The name of the estimator class. We pass estimator name
+          here because there are some wrapped methods in Pipeline that don't
+          have ``__self__`` to access the class name.
+
+    Returns:
+        The passed function wrapped in a span.
+    """
+    logger.debug("Instrumenting: %s.%s", estimator_name, func.__name__)
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        with get_tracer(__name__, __version__).start_as_current_span(
+            name="{cls}.{func}".format(cls=estimator_name, func=func.__name__)
+        ):
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+# Methods on which spans should be applied.
+DEFAULT_METHODS = ["fit", "transform", "predict", "predict_proba"]
+
+# Classes and their attributes which contain a list of tupled estimators
+# through which we should walk recursively for estimators.
+DEFAULT_NAMEDTUPLE_ATTRIBS = {
+    Pipeline: ["steps"],
+    FeatureUnion: ["transformer_list"],
+}
+
+# Classes and their attributes which contain an estimator or sequence of
+# estimators through which we should walk recursively for estimators.
+DEFAULT_ATTRIBS = {}
+
+
+class SklearnInstrumentor:
```

Comment:
I believe we want all instrumentors to derive from [BaseInstrumentor](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-instrumentation/src/opentelemetry/instrumentation/instrumentor.py#L26). Users should instrument manually with the `instrumentor().instrument()` method.

Reply:
This instrumentation doesn't actually instrument the library, it instruments a trained model. It's a bit different from the other instrumentation packages in that regard, and I hadn't intended on providing any auto-instrumentation. There are a few reasons for this:

1. In an ML application with many models, be able to instrument different machine learning models differently (or some not at all) by instantiating different instrumentors.
2. More easily configure which estimators in which models are instrumented.
3. Sklearn has an extensible API and there are many other packages like sklearn that we might also want to instrument. A user might also want to instrument their own home grown estimators. We can do this more easily at the model than the library imho.
4. There are some estimators, like decision trees that we probably don't want to instrument if they are part of an ensemble model of hundreds or thousands of trees; it could generate unwanted overhead at runtime.
5. Instrumenting the library would mean walking the module hierarchy and attempting to patch every class's API with instrumentation. An alternative would be to keep a master list of classes on which to instrument methods which may not exist on each class, but this could break easily on sklearn releases.

If this is an acceptable edge case I think that it might not need to inherit from BaseInstrumentor, but maybe the lack of auto-instrumentation can be made more explicit in documentation. Otherwise, I think we might want to discuss further.

Reply:
It seems like the use case is similar to the [Flask instrumentor](https://github.com/open-telemetry/opentelemetry-python/blob/master/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L173) in which we provide a mechanism for users to instrument individual flask applications, or the whole library itself. Could we maybe do something like that? The reason I'm leaning towards inheriting from `BaseInstrumentor` is not only for auto-instrumentation, but for when we want to have some logic that applies to all instrumentations, we can be confident that it will cover all the packages we want if they inherit from them (e.g. [metric collection](https://github.com/open-telemetry/opentelemetry-python/pull/1116)).

> Instrumenting the library would mean walking the module hierarchy and attempting to patch every class's API with instrumentation.

However it sounds like patching the entire library is difficult? I'm not too familiar with sklearn's api, but is there not a base class that each model type inherits from?

Reply:
It might be easier than I thought. I'll take a look at it again.

Reply:
It wasn't as easy as I thought but the latest push should support both package and model instrumentation.

Reply:
The challenge comes from [this decorator](https://scikit-learn.org/stable/modules/generated/sklearn.utils.metaestimators.if_delegate_has_method.html) which delegates attributes to sub-estimators. I've gotten this to work for instances of estimators but the current iteration doesn't quite work on patching classes. So there is still a bit of work on the package instrumentation side.